### ISSUE TITLE
Conformance to Hashable is broken for objects without primary key

### ISSUE THREAD

**Issue by bdash** [realm/realm-cocoa#4705]
Classes deriving from `Object` appear to automatically inherit conformance to `Hashable` due to having `NSObject` as a base class. The implementation of `hashValue` comes from `-[RLMObjectBase hash]`, by way of `NSObject`'s `Hashable` conformance delegating to the object's `hash` property. `-[RLMObjectBase hash]` isn't correct with respect to `-[RLMObjectBase isEqual:]` for objects that don't have a primary key declared, `-hash` will return different values for two objects that compare as equal, which violates the assumptions about how `Hashable` objects are used.

**Comment by tgoyne** [comment]
`-[RLMObjectBase isEqual:]` is only overridden for objects which have a primary key for specifically this reason.

**Comment by bdash** [comment]
Hmm, yeah, I overlooked that. I guess that means the behavior we provide is correct, but (unavoidably) not super useful for objects without a primary key.

**Comment by bdash** [comment]
I was wrong about the cause, but this really _is_ broken at present. `Object` overrides `isEqual` and unconditionally calls `RLMObjectBaseAreEqual` rather than only calling it if the class has a primary key as `-[RLMObjectBase isEqual:]` does.

**Comment by bdash** [comment]
Removing the override of `Object.isEqual(_:)` introduces many test failures as we appear to be relying on the same object retrieved multiple times from the same Realm comparing as equal.

**Comment by CraigSiemens** [comment]
I just ran into this issue. I'm putting elements into a Set and sometimes the set will end up containing multiple instances of the same object because they have different hashValues even though `==` is true.

**Comment by li-kai** [comment]
Overriding hashvalue seems to do the trick without having to declare a primary key.
```swift
override var hashValue: Int {
    return x.hashValue
}
```

**Comment by austinzheng** [comment]
Fixed by #5080; merged to `master-3.0`.

**Comment by maroony** [comment]
Maybe this is broken again in 4.3.0 (see also: [Stackoverflow](https://stackoverflow.com/questions/54111153/realmswift-how-to-override-hash-function-when-migrating-pre-existing-class-to)). 

The only thing you could override is 

```swift
  override var hash: Int {
      return title.hashValue
  }
```

but this won't help if you use this class as a Key-Class in a dictionary.
